sktime/utils/_testing/estimator_checks.py

```diff
@@ -430,6 +476,38 @@ def check_methods_do_not_change_state(Estimator):
             ), f"Estimator: {estimator} changes __dict__ during {method}"
 
 
+def check_methods_have_no_side_effects(Estimator):
+    # Check that calling methods has no side effects on args
+
+    if not isclass(Estimator):
+        Estimator = type(Estimator)
+
+    estimator = _construct_instance(Estimator)
+
+    set_random_state(estimator)
+
+    # Fit for the first time
+    fit_args = _make_args(estimator, "fit")
+    old_fit_args = deepcopy(fit_args)
+    estimator.fit(*fit_args)
+
+    assert dict_equals(
```

Comment:
`args` are tuples here, does the built-in tuple comparison not work, e.g. `a == b`?

Reply:
yes, just noticed that, see commit below

Reply:
no, comparison does not work as soon as you have types inside that use custom equality checks which are not defaulted to by `==`

Reply:
for instance `np.array` does this element-wise, and `pandas` types return errors so you have to use `equals`